### PR TITLE
refactor(storage): do not aware of keyspace in state table

### LINKED ISSUE #3212: storage: make state table not aware of the keyspace prefix
I still think the state table should be built based on Keyspace, and none of the interfaces should be aware of the keyspace's prefix. This will definitely make the implementation much clearer and bug-freer. 🥹

I'd like to do this refactor while reviewing the state table these days.

_Originally posted by @BugenZhao in https://github.com/singularity-data/risingwave/issues/3191#issuecomment-1154700422_

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>

## What's changed and what's your intention?

With the `prefixed_range` introduced in #3217, we can make the cell-based table & state table not aware of keyspace prefix with ease.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- Close #3212

### CODE REVIEW

**Review by xxchan — COMMENTED** [review]
make `Keyspace::prefixed_key` private?

File: src/storage/src/table/cell_based_table.rs [hunk 1]
```diff
@@ -382,22 +376,20 @@ impl<S: StateStore> CellBasedTable<S> {
...
+                } else if pk_prefix.size() == 0 {
+                    Unbounded
                 } else {
-                    Excluded(if pk_prefix.size() == 0 {
-                        next_key(self.keyspace.key())
-                    } else {
-                        self.keyspace.prefixed_key(&next_key(&serialized_pk_prefix))
-                    })
+                    Excluded(next_key(&serialized_pk_prefix))
                 }
```
Comment (xxchan) [inline src/storage/src/table/cell_based_table.rs:393]: Should it be
```rust
if pk_prefix.size() == 0 {
	Unbounded
} else {
	if is_start_bound {
		Included(serialized_pk_prefix)
	} else {
		Excluded(next_key(&serialized_pk_prefix))
	}
}

```
?

**Comment by wcy-fdu** [comment]
May conflict with https://github.com/singularity-data/risingwave/pull/3238.

File: src/storage/src/keyspace.rs [hunk 2]
```diff
@@ -88,63 +87,63 @@ impl<S: StateStore> Keyspace<S> {
...
         epoch: u64,
     ) -> StorageResult<Vec<(Bytes, Bytes)>> {
-        let range = range_of_prefix(&self.prefix);
+        self.scan_with_range::<_, &[u8]>(.., limit, epoch).await
+    }
+
+    /// Scans `limit` keys from the given `range` in this keyspace and get their values.
+    /// Note that the `range` should not be prepended with the prefix of this keyspace.
+    /// If `limit` is None, all keys of the given prefix will be scanned.
+    /// The returned values are based on a snapshot corresponding to the given `epoch`.
```
Comment (xxchan) [inline src/storage/src/keyspace.rs:109]:
```suggestion
    /// Scans `limit` keys from the given `range` in this keyspace and get their values.
    /// If `limit` is None, all keys of the given prefix will be scanned.
    /// The returned values are based on a snapshot corresponding to the given `epoch`.
    ///
    /// **Note**: the `range` should not be prepended with the prefix of this keyspace.
```
Make the note more noticable and the remaining the same as `scan` 😁

File: src/storage/src/keyspace.rs [hunk 3]
```diff
@@ -88,63 +87,63 @@ impl<S: StateStore> Keyspace<S> {
...
-            iter,
-            prefix_len: self.prefix.len(),
-        };
-        Ok(strip_prefix_iterator)
+        self.iter_with_range::<_, &[u8]>(.., epoch).await
     }
 
+    /// Gets an iterator of the given `range` in this keyspace.
+    /// Note that the `range` should not be prepended with the prefix of this keyspace.
+    /// The returned iterator will iterate data from a snapshot corresponding to the given `epoch`.
```
Comment (xxchan) [inline src/storage/src/keyspace.rs:136]: ditto
```suggestion
    /// Gets an iterator of the given `range` in this keyspace.
    /// The returned iterator will iterate data from a snapshot corresponding to the given `epoch`.
    /// **Note**: `range` should not be prepended with the prefix of this keyspace.
```

**Review by skyzh — APPROVED** [review]

**Review by lmatz — APPROVED** [review]
🙇‍♂️

**Review by xxchan — APPROVED** [review]
look much better!

**Review by wcy-fdu — APPROVED** [review]

**Comment by BugenZhao** [comment]
> make `Keyspace::prefixed_key` private?

We're still using it in `write_batch.rs`. Maybe we can make this a submodule of `keyspace` so that we can make it private.

**Pull request merged**